torch/csrc/jit/codegen/cuda/lower_unroll.cpp

```diff
@@ -109,7 +109,81 @@ void UnrollPass::handle(Expr* expr) {
+    auto is_ld_matrix_op = [](Expr* expr) {
+      if (auto uop = dynamic_cast<UnaryOp*>(expr)) {
+        if (uop->getUnaryOpType() == UnaryOpType::LD_MATRIX ||
+            uop->getUnaryOpType() == UnaryOpType::LD_MATRIXT) {
+          return true;
+        }
+      }
+      return false;
+    };
+
+    // Check if the given op is an async copy
+    auto is_cpasync_op = [](Expr* expr) {
+      if (auto uop = dynamic_cast<UnaryOp*>(expr)) {
+        if (uop->getUnaryOpType() == UnaryOpType::CP_ASYNC) {
+          return true;
+        }
+      }
+      return false;
+    };
+
+    bool is_ld_matrix = false;
+    bool is_ld_matrix_producer = false;
+    if (is_ld_matrix_op(expr)) {
+      // Note & TODO: currently cannot support predicated ldmatrix since it'd
+      //  need to make sure all threads in a warp evaluate the predicate
+      //  to the same value.
+      // For now asserting that the predicate for ldmatrix and its producer
+      //  can be omitted. Would need to build out support for warp level
+      //  predicates if we want vectorization predicates to be active on
+      //  these ops.
+      TORCH_INTERNAL_ASSERT(
+          GpuLower::current()->predicateElimination().canOmitPredicate(expr),
+          "Vectorize: unsupported predicate for warp ops");
+      is_ld_matrix = true;
+    }
+
+    // Note & TODO: This part is a WAR that disables any predicated
+    //  producer of ldmatrix consumers. Ldmatrix currently cannot
+    //    be predicated due to no support in warp-uniform predicates.
+    //  A predicated producer will result in ldmatrix propagating
+    //   data outside of the producer's valid region.
+    //  Except for vectorized global loads which should be properly
+    //  initialized outside of the predicated region.
+    if (!std::any_of(
+            expr->inputs().begin(), expr->inputs().end(), [](Val* val) {
+              return val->isA<TensorView>() &&
+                  val->as<TensorView>()->getMemoryType() == MemoryType::Global;
+            })) {
+      for (auto out_tv : ir_utils::filterByType<TensorView>(expr->outputs())) {
+        if (is_cpasync_op(out_tv->definition())) {
+          // cp async op has initialization outside of the predicate protected
+          //  boundary so no need to ensure predicate removal.
+          break;
+        }
+
+        for (auto use : expr->fusion()->unordered_uses(out_tv)) {
+          if (is_ld_matrix_op(use)) {
+            TORCH_INTERNAL_ASSERT(
+                GpuLower::current()->predicateElimination().canOmitPredicate(
+                    expr),
+                "Vectorize: producers of ldmatrix cannot be predicated except for global loads:",
+                expr);
+            is_ld_matrix_producer = true;
+            break;
+          }
+        }
+      }
+    }
+
+    const bool omit_vector_predicate_for_ldmatrix =
+        is_ld_matrix || is_ld_matrix_producer;
+
+    if (!unswitched_loop_ && !omit_vector_predicate_for_ldmatrix &&
```

Comment:
This part is trying to remove the vectorization predicate that was inserted in lower_unroll pass, since predicate_removal pass doesn't remove vectorization predicates. It seems to need a cleanup on a larger scale. We could revisit this.

Reply:
Can you show an example? Vectorize predicates use `UnswitchPredicate`, where predicate elimination is at least looked up.

Reply:
Never mind. None of the changes in this file is needed anymore. Will clear it.

Reply:
Just found this was actually due to a bug in predicate elimination pass. Tracking in  #1641.

Reply:
Patched in #1642